### PR TITLE
Improve C++ SDK perf 5x by respecting CMAKE_BUILD_TYPE and enabling mimalloc

### PR DESCRIPTION
**Pull request by jleibs**
### What
- Resolves: https://github.com/rerun-io/rerun/issues/4093

Arrow was default to a Debug build regardless of what CMAKE_BUILD_TYPE was specified.

Baseline:
```
$ time RERUN=on ./build/rerun_vrs_example sample.vrs
real    0m15.105s
user    0m16.409s
sys     0m2.981s
```

`CMAKE_BUILD_TYPE=RelWithDebInfo`
```
$ time RERUN=on ./build/rerun_vrs_example sample.vrs
real    0m4.659s
user    0m5.259s
sys     0m2.257s
```

`-DARROW_MIMALLOC=ON`
```
$ time RERUN=on ./build/rerun_vrs_example sample.vrs
real    0m3.477s
user    0m5.015s
sys     0m1.291s
```

### Test
* [x] Linux
* [x] Mac
* [x] Windows, rerun build & test only

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)
* [x] I have tested [demo.rerun.io](https://demo.rerun.io/pr/4094) (if applicable)
* [x] The PR title and labels are set such as to maximize their usefulness for the next release's CHANGELOG

- [PR Build Summary](https://build.rerun.io/pr/4094)
- [Docs preview](https://rerun.io/preview/73f97330c4d4e8cbac82e7e0c0ea7e9d24fcd3c7/docs) 
- [Examples preview](https://rerun.io/preview/73f97330c4d4e8cbac82e7e0c0ea7e9d24fcd3c7/examples) 
- [Recent benchmark results](https://ref.rerun.io/dev/bench/)
- [Wasm size tracking](https://ref.rerun.io/dev/sizes/)

### CODE REVIEW

**Review by teh-cmc — APPROVED** [review]

File: rerun_cpp/CMakeLists.txt [hunk 1]
```diff
@@ -168,18 +174,20 @@ if(RERUN_DOWNLOAD_AND_BUILD_ARROW)
...
         --preset ninja-debug-minimal
-        -DARROW_IPC=ON
+        -DARROW_BOOST_USE_SHARED=OFF
         -DARROW_BUILD_SHARED=${ARROW_BUILD_SHARED}
         -DARROW_BUILD_STATIC=${ARROW_BUILD_STATIC}
-        -DCMAKE_INSTALL_PREFIX=${ARROW_DOWNLOAD_PATH}
+        -DARROW_CXXFLAGS=${DARROW_CXXFLAGS}
+        -DARROW_IPC=ON
+        -DARROW_JEMALLOC=OFF
+        -DARROW_MIMALLOC=ON
```
Comment (emilk) [inline rerun_cpp/CMakeLists.txt:183]: is mimalloc automatically built from source, or will this cause new linking issues?

**Comment by Wumpf** [comment]
Let's make sure this works on Mac & Windows before mergning
I remember disabling jemalloc because of build issues somewhere (don't know which one) and turning it off helped (didn't think much of perf impact at that point ofc :( ). Still opens the question though whether we should use jemalloc over mimalloc or not.

File: rerun_cpp/CMakeLists.txt (hunk 1 above)
Comment (jleibs) [reply rerun_cpp/CMakeLists.txt:183]: Yeah, it does it's own fetch + build, the same way we do with arrow:
```
[6/145] Performing download step (download, verify and extract) for 'mimalloc_ep'
-- Downloading...
   dst='/home/jleibs/cpp-example-vrs/build/arrow/src/arrow_cpp-build/mimalloc_ep-prefix/src/v2.0.6.tar.gz'
   timeout='none'
   inactivity timeout='none'
-- Using src='https://github.com/microsoft/mimalloc/archive/v2.0.6.tar.gz'
-- verifying file...
       file='/home/jleibs/cpp-example-vrs/build/arrow/src/arrow_cpp-build/mimalloc_ep-prefix/src/v2.0.6.tar.gz'
-- Downloading... done
-- extracting...
     src='/home/jleibs/cpp-example-vrs/build/arrow/src/arrow_cpp-build/mimalloc_ep-prefix/src/v2.0.6.tar.gz'
     dst='/home/jleibs/cpp-example-vrs/build/arrow/src/arrow_cpp-build/mimalloc_ep-prefix/src/mimalloc_ep'
```

**Comment by emilk** [comment]
`just cpp-test` works on Mac

File: rerun_cpp/CMakeLists.txt [hunk 2]
```diff
@@ -168,18 +174,20 @@ if(RERUN_DOWNLOAD_AND_BUILD_ARROW)
...
         -DARROW_USE_TSAN=OFF
         -DARROW_USE_UBSAN=OFF
-        -DARROW_JEMALLOC=OFF
-        -Dxsimd_SOURCE=BUNDLED
         -DBOOST_SOURCE=BUNDLED
-        -DARROW_BOOST_USE_SHARED=OFF
-        -DARROW_CXXFLAGS=${DARROW_CXXFLAGS}
+        -DCMAKE_BUILD_TYPE=${ARROW_CMAKE_BUILD_TYPE}
+        -DCMAKE_INSTALL_PREFIX=${ARROW_DOWNLOAD_PATH}
+        -Dxsimd_SOURCE=BUNDLED
```
Comment (emilk) [inline rerun_cpp/CMakeLists.txt:190]: We should sort these defines if their order doesn't matter

**Review by emilk — APPROVED** [review]

**Comment by Wumpf** [comment]
oh wow. CRLF strikes again. my settings are broken

**Review by Wumpf — APPROVED** [review]
fixed for windows now, didn't run benchmarks though, only made sure build all & tests work

**Comment by emilk** [comment]
tip: there are global git settings you can set regarding CRLF

File: rerun_cpp/CMakeLists.txt [hunk 3]
```diff
@@ -159,10 +160,18 @@ if(RERUN_DOWNLOAD_AND_BUILD_ARROW)
         set(DARROW_CXXFLAGS "")
     endif()
 
-    if(NOT CMAKE_BUILD_TYPE)
-        set(ARROW_CMAKE_BUILD_TYPE Release)
+    # Workaround for https://stackoverflow.com/questions/35519538/visual-studio-code-tab-key-does-not-insert-a-tab
```
Comment (emilk) [inline rerun_cpp/CMakeLists.txt:163]: I'm pretty sure this is the wrong link

**Comment by Wumpf** [comment]
> tip: there are global git settings you can set regarding CRLF

I know, that's the strange thing, I set `autocrlf = false` on my global .gitconfig and I still got this. Must be something else in VSCode that caused this because changing the endings and commting that worked fine 😕

File: rerun_cpp/src/rerun/demo_utils.hpp [hunk 4]
```diff
@@ -30,7 +30,7 @@ namespace rerun {
         inline std::vector<T> linspace(T start, T end, size_t num) {
             std::vector<T> linspaced(num);
             std::generate(linspaced.begin(), linspaced.end(), [&, i = 0]() mutable {
-                return start + static_cast<T>(i++) * (end - start) / static_cast<T>(num - 1);
+                return static_cast<T>(start + static_cast<T>(i++) * (end - start) / static_cast<T>(num - 1));
```
Comment (emilk) [inline rerun_cpp/src/rerun/demo_utils.hpp:33]: this makes no sense to me - isn't this casting `T` to `T`? Are you sure this is needed?

**Review by emilk — APPROVED** [review]

File: rerun_cpp/src/rerun/demo_utils.hpp (hunk 4 above)
Comment (Wumpf) [reply rerun_cpp/src/rerun/demo_utils.hpp:33]: Yeah it's this weird quirk of C++ where `char + char == int`

Comment (Wumpf) [reply rerun_cpp/src/rerun/demo_utils.hpp:33]: (you know _that_ weird quirk. Not like the others)
Bites me every time!

https://godbolt.org/z/4W8xYnqKq

actually only works with unsigned char, which is also the case I got a warning in (uint8_t)

Comment (emilk) [reply rerun_cpp/src/rerun/demo_utils.hpp:33]: God damn you C++

File: rerun_cpp/CMakeLists.txt (hunk 2 above)
Comment (jleibs) [reply rerun_cpp/CMakeLists.txt]: This PR sorts them.

**Comment by jleibs** [comment]
@emilk can you re-test on mac with the latest changes?

**Pull request merged**